gadopt/level_set_tools.py

```diff
@@ -713,32 +715,128 @@ class was initialised.
+    level_set: fd.Function, material_area: float, entrainment_height: float
+) -> float:
+    """Calculates the entrainment diagnostic.
 
-    Determines the proportion of a material that is located above a given height.
+    Determines the proportion of a material located above a given height.
 
     Args:
-        level_set:
-          A level-set Firedrake function.
-        material_area:
-          An integer or a float representing the total area occupied by a material.
-        entrainment_height:
-          An integer or a float representing the height above which the entrainment
-          diagnostic is determined.
+      level_set:
+        A Firedrake function for the level-set field
+      material_area:
+        A float representing the total area occupied by the target material
+      entrainment_height:
+        A float representing the height above which to calculate entrainment
 
     Returns:
-        A float corresponding to the calculated entrainment diagnostic.
+      A float corresponding to the material fraction above the target height
     """
-    mesh_coords = fd.SpatialCoordinate(level_set.function_space().mesh())
+    mesh_coords = node_coordinates(level_set)
     target_region = mesh_coords[1] >= entrainment_height
     material_entrained = fd.conditional(level_set < 0.5, 1, 0)
+    is_entrained = fd.conditional(target_region, material_entrained, 0)
 
-    return (
-        fd.assemble(fd.conditional(target_region, material_entrained, 0) * fd.dx)
-        / material_area
-    )
+    return fd.assemble(is_entrained * fd.dx) / material_area
+
+
+def min_max_height(
+    level_set: fd.Function, epsilon: float | fd.Function, *, side: int, mode: str
+) -> float:
+    """Calculates the maximum or minimum height of a material interface.
+
+    Args:
+      level_set:
+        A Firedrake function for the level set field
+      epsilon:
+        A float or Firedrake function denoting the thickness of the material interface
+      side:
+        An integer (0 or 1) indicating the level-set side of the target material
+      mode:
+        A string ("min" or "max") specifying which extremum height is sought
+
+    Returns:
+      A float corresponding to the material interface extremum height
+    """
+
+    match side:
+        case 0:
+            comparison = operator.le
+        case 1:
+            comparison = operator.ge
+        case _:
+            raise ValueError("'side' must be 0 or 1.")
+
+    match mode:
+        case "min":
+            extremum = np.min
+            ls_arg_extremum = np.argmax
+            irrelevant_data = np.inf
+            mpi_comparison = MPI.MIN
+        case "max":
+            extremum = np.max
+            ls_arg_extremum = np.argmin
+            irrelevant_data = -np.inf
+            mpi_comparison = MPI.MAX
+        case _:
+            raise ValueError("'mode' must be 'min' or 'max'.")
+
+    mesh = level_set.ufl_domain()
+    if not mesh.cartesian:
```

Comment:
Can we extend this to non Cartesian, based on radius?

Reply:
The level-set implementation has been designed with rectangular and cuboid domains in mind; it has not been tested with any spherical or cylindrical domains. I would prefer having this extension as a separate PR.